File: document_processing/pipeline_modules/page_registration/line_refine.py
"""Blur-tolerant refinement of a rectified page by its own straight structures.

After the page is warped into its frame (from a Borders quad or a template
homography) every horizontal structure of a flat page - field rules, text
baselines, the MRZ, the red band - must be exactly horizontal and every
vertical one (photo frame, series/number column, page edges) exactly
vertical. Whatever residual tilt is left tells how the warp is off:

- the same tilt everywhere      -> rotation (or shear, if verticals disagree);
- tilt changing with y          -> horizontal lines converge: perspective p1;
- vertical tilt changing with x -> vertical lines converge: perspective p2.

Evidence comes from two sources so that it survives blur, which kills keypoint
matching first: LSD line segments (plenty on a sharp page, few on a blurred
one) and projection-profile tilts of horizontal bands and vertical strips
(text lines and rules make a sharp profile even when badly blurred; measured
on frames where SIFT found 2-5 inliers). A 4-parameter model (rotation, shear,
two perspective terms) is fitted to the angles with a robust loss; parameters
without evidence are pinned to zero by a weak prior. The correction is applied
only if it is small and actually reduces the residual - otherwise the page is
left alone. No template, no weights, deterministic.
"""
from typing import List, Optional, Tuple

import cv2
import numpy as np
from scipy.optimize import least_squares

LSD_MIN_LEN_FRAC = 0.04       # segment length, fraction of page width
LSD_MAX_WEIGHT = 3.0          # a segment weighs len / (0.1 * W), capped
ANGLE_TOL_DEG = 12.0          # |angle| below this is 'horizontal', near 90 'vertical'
PROFILE_SCALE = 0.5
PROFILE_COARSE = np.arange(-6.0, 6.01, 1.0)
PROFILE_FINE_STEP = 0.25
PROFILE_MIN_PEAK = 1.15       # best/median score ratio for a band's tilt to count
PROFILE_WEIGHT = 3.0
MIN_HORIZ_WEIGHT = 4.0        # total weight of horizontal evidence needed to act
MAX_ROT_DEG = 5.0
MAX_CORNER_SHIFT_FRAC = 0.08  # of the page width
MIN_GAIN = 0.25               # weighted median |residual| must drop by this fraction
MIN_RESIDUAL_DEG = 0.3        # ... and be worth correcting to begin with
PRIOR_SCALE = (0.03, 0.05, 0.03)  # shear / p1 / p2 of this size cost one degree of residual:
                              # shear and p2 need vertical evidence, which is scarce, so
                              # they are held harder than p1
IRLS_ITERS = 3
IRLS_SCALE_DEG = 1.0          # Cauchy scale of the measurement weights
VERT_MIN_LEN_FRAC = 0.15      # a 'vertical' segment must be this long (of the page height):
                              # page edges, the photo frame, the ornament band - not hair,
                              # not a stroke of a letter (short verticals set p2 to junk)
BLOB_MAX_FRAC = 0.2           # ink components taller (bands) / wider (strips) than this
                              # fraction of the region are not text: dropped from the profile
N_BANDS = 5                   # overlapping bands / strips, each a third of the page

_LSD = None


def _lsd():
    global _LSD
    if _LSD is None:
        _LSD = cv2.createLineSegmentDetector(cv2.LSD_REFINE_STD)
    return _LSD


def _segments(gray: np.ndarray, min_len: float) -> np.ndarray:
    lines = _lsd().detect(gray)[0]
    if lines is None:
        return np.zeros((0, 4), np.float64)
    lines = lines.reshape(-1, 4).astype(np.float64)
    L = np.hypot(lines[:, 2] - lines[:, 0], lines[:, 3] - lines[:, 1])
    return lines[L >= min_len]


def _seg_angle(segs: np.ndarray) -> np.ndarray:
    ang = np.degrees(np.arctan2(segs[:, 3] - segs[:, 1], segs[:, 2] - segs[:, 0]))
    return (ang + 90.0) % 180.0 - 90.0          # (-90, 90]


def _profile_tilt(region: np.ndarray, axis: int) -> Tuple[float, float]:
    """Tilt (deg) that makes the projection profile along ``axis`` sharpest,
    and the peak ratio (best score / median score). axis=1: row sums, i.e.
    horizontal text lines; axis=0: column sums, vertical structures."""
    small = region
    if PROFILE_SCALE != 1.0:
        small = cv2.resize(region, None, fx=PROFILE_SCALE, fy=PROFILE_SCALE, interpolation=cv2.INTER_AREA)
    # ink = 1, paper = 0 (Otsu, as the deskewer does) so the empty corners a
    # rotation leaves look like paper; the profile is still normalised by the
    # number of valid pixels per row so a cut corner cannot tilt the score
    _, binary = cv2.threshold(small, 0, 255, cv2.THRESH_BINARY_INV + cv2.THRESH_OTSU)
    # keep text-sized ink only: a photo, a stamp, a thumb or the dark table
    # beyond a page edge is one big blob whose outline - not the text lines -
    # would otherwise own the profile (measured: bands disagreeing in sign)
    h, w = binary.shape
    n, lab, st, _ = cv2.connectedComponentsWithStats(binary, connectivity=8)
    if n > 1:
        big = (st[:, cv2.CC_STAT_HEIGHT] > BLOB_MAX_FRAC * h) if axis == 1 \
            else (st[:, cv2.CC_STAT_WIDTH] > BLOB_MAX_FRAC * w)
        big[0] = False
        if big.any():
            binary[big[lab]] = 0
    binary = binary.astype(np.float32) / 255.0
    valid = np.ones_like(binary)
    centre = (w / 2.0, h / 2.0)

    def score(angles):
        out = np.empty(len(angles))
        for i, a in enumerate(angles):
            M = cv2.getRotationMatrix2D(centre, float(a), 1.0)
            rot = cv2.warpAffine(binary, M, (w, h), flags=cv2.INTER_NEAREST, borderValue=0)
            cnt = cv2.warpAffine(valid, M, (w, h), flags=cv2.INTER_NEAREST, borderValue=0).sum(axis=axis)
            ink = rot.sum(axis=axis)
            ok = cnt >= 0.6 * cnt.max()
            prof = ink[ok] / np.maximum(cnt[ok], 1.0)
            out[i] = float(prof.var()) if len(prof) > 2 else 0.0
        return out

    coarse = score(PROFILE_COARSE)
    ib = int(np.argmax(coarse))
    if ib == 0 or ib == len(PROFILE_COARSE) - 1 or coarse[ib] <= 0:
        return 0.0, 0.0          # no peak inside the range: no evidence
    fine = np.arange(PROFILE_COARSE[ib] - 1.0, PROFILE_COARSE[ib] + 1.001, PROFILE_FINE_STEP)
    fs = score(fine)
    jb = int(np.argmax(fs))
    best = float(fine[jb])
    if 0 < jb < len(fine) - 1:                   # parabolic interpolation
        y0, y1, y2 = fs[jb - 1], fs[jb], fs[jb + 1]
        den = y0 - 2 * y1 + y2
        if den < 0:
            best += PROFILE_FINE_STEP * 0.5 * (y0 - y2) / den
    med = float(np.median(coarse))
    ratio = float(fs[jb] / med) if med > 0 else 0.0
    # cv2 rotates counter-clockwise for a positive angle, which straightens a
    # line running down to the right, i.e. one with a POSITIVE image angle
    # (atan2 with y down) - so the best angle is the tilt of the structure
    # itself, in the same convention as the LSD segments (pinned by a test)
    return best, ratio


def measure(gray: np.ndarray, inset: int = 0) -> np.ndarray:
    """Straightness evidence of a rectified page.

    Returns rows (x1, y1, x2, y2, kind, weight, source): kind 0 = should be
    horizontal, 1 = should be vertical; source 0 = LSD segment, 1 = band
    profile, 2 = strip profile. ``inset`` excludes the cushion around the page
    from the profile regions (segments are taken everywhere: a page edge
    visible in the cushion is evidence too)."""
    H, W = gray.shape[:2]
    rows: List[list] = []
    segs = _segments(gray, LSD_MIN_LEN_FRAC * W)
    if len(segs):
        ang = _seg_angle(segs)
        L = np.hypot(segs[:, 2] - segs[:, 0], segs[:, 3] - segs[:, 1])
        wgt = np.minimum(L / (0.1 * W), LSD_MAX_WEIGHT)
        for s, a, w, l in zip(segs, ang, wgt, L):
            if abs(a) < ANGLE_TOL_DEG:
                rows.append([*s, 0, w, 0])
            elif abs(abs(a) - 90.0) < ANGLE_TOL_DEG and l >= VERT_MIN_LEN_FRAC * H:
                rows.append([*s, 1, w, 0])
    x0, y0, x1, y1 = inset, inset, W - inset, H - inset
    if x1 - x0 < 60 or y1 - y0 < 60:
        return np.array(rows, np.float64).reshape(-1, 7)
    # overlapping horizontal bands: tilt of the text lines at several heights
    bh = (y1 - y0) / 3.0
    for yb in np.linspace(y0, y1 - bh, N_BANDS):
        band = gray[int(yb):int(yb + bh), x0:x1]
        tilt, ratio = _profile_tilt(band, axis=1)
        if ratio >= PROFILE_MIN_PEAK:
            yc = yb + 0.5 * bh
            half = 0.5 * (x1 - x0)
            dy = np.tan(np.radians(tilt)) * half
            rows.append([x0, yc - dy, x1, yc + dy, 0, PROFILE_WEIGHT * min(1.0, ratio - 1.0), 1])
    # vertical strips (column profiles) were tried and dropped: a passport page
    # has almost no vertical text structure, and the strip tilts came out as
    # junk (measured -5 deg on straight pages) that only fed p2 and shear
    return np.array(rows, np.float64).reshape(-1, 7)


def _model(params, W, H) -> np.ndarray:
    """Pixel homography for (rotation rad, shear, p1, p2) about the page centre,
    perspective terms in units of the half page width."""
    th, s, p1, p2 = params
    k = 2.0 / W
    T = np.array([[k, 0, -1.0], [0, k, -H / W], [0, 0, 1.0]])
    R = np.array([[np.cos(th), -np.sin(th), 0], [np.sin(th), np.cos(th), 0], [0, 0, 1.0]])
    S = np.array([[1.0, s, 0], [0, 1.0, 0], [0, 0, 1.0]])
    P = np.array([[1.0, 0, 0], [0, 1.0, 0], [p1, p2, 1.0]])
    return np.linalg.inv(T) @ P @ R @ S @ T


def _angles_after(Hm, meas) -> np.ndarray:
    p0 = cv2.perspectiveTransform(meas[:, 0:2].reshape(1, -1, 2), Hm).reshape(-1, 2)
    p1 = cv2.perspectiveTransform(meas[:, 2:4].reshape(1, -1, 2), Hm).reshape(-1, 2)
    ang = np.degrees(np.arctan2(p1[:, 1] - p0[:, 1], p1[:, 0] - p0[:, 0]))
    ang = (ang + 90.0) % 180.0 - 90.0
    dev = np.where(meas[:, 4] == 0, ang, np.abs(ang) - 90.0)
    return dev


def _residuals(params, meas, W, H, wgt):
    dev = _angles_after(_model(params, W, H), meas) * np.sqrt(wgt)
    prior = np.array(params[1:]) / np.array(PRIOR_SCALE)
    return np.concatenate([dev, prior])


def _fit(meas, W, H):
    """Robust fit of the 4 parameters: iteratively reweighted least squares
    with a Cauchy weight on every measurement (an outlying band or a segment
    on a hair line stops mattering) while the prior on the parameters stays
    quadratic - a plain robust loss would discount a large prior residual
    exactly like an outlier and let a single junk vertical segment set p2."""
    x = np.zeros(4)
    wgt = meas[:, 5].copy()
    for _ in range(IRLS_ITERS):
        sol = least_squares(_residuals, x, args=(meas, W, H, wgt), max_nfev=100)
        x = sol.x
        r = _angles_after(_model(x, W, H), meas)
        wgt = meas[:, 5] / (1.0 + (r / IRLS_SCALE_DEG) ** 2)
    return x


def _wmedian(values: np.ndarray, weights: np.ndarray) -> float:
    if len(values) == 0:
        return 0.0
    order = np.argsort(values)
    v, w = values[order], weights[order]
    c = np.cumsum(w)
    return float(v[np.searchsorted(c, 0.5 * c[-1])])


def refine_by_lines(gray: np.ndarray, inset: int = 0,
                    min_residual_deg: float = MIN_RESIDUAL_DEG) -> Tuple[Optional[np.ndarray], dict]:
    """Homography (pixel, src->dst for cv2.warpPerspective) that straightens a
    rectified page by its own lines, or None when there is not enough evidence
    or the correction is not warranted. ``info`` carries the evidence counts,
    the residual before/after (weighted median |deg|) and the parameters."""
    H, W = gray.shape[:2]
    meas = measure(gray, inset)
    info = {'n_seg': int((meas[:, 6] == 0).sum()) if len(meas) else 0,
            'n_band': int((meas[:, 6] == 1).sum()) if len(meas) else 0,
            'n_strip': int((meas[:, 6] == 2).sum()) if len(meas) else 0,
            'applied': False}
    if len(meas) == 0:
        info['reason'] = 'no evidence'
        return None, info
    horiz_w = float(meas[meas[:, 4] == 0, 5].sum())
    info['horiz_weight'] = round(horiz_w, 2)
    if horiz_w < MIN_HORIZ_WEIGHT:
        info['reason'] = 'too little horizontal evidence'
        return None, info
    before = np.abs(_angles_after(np.eye(3), meas))
    info['before'] = round(_wmedian(before, meas[:, 5]), 2)
    x = _fit(meas, W, H)
    Hm = _model(x, W, H)
    after = np.abs(_angles_after(Hm, meas))
    info['after'] = round(_wmedian(after, meas[:, 5]), 2)
    info['params'] = {'rot_deg': round(float(np.degrees(x[0])), 2), 'shear': round(float(x[1]), 4),
                      'p1': round(float(x[2]), 4), 'p2': round(float(x[3]), 4)}
    if info['before'] < min_residual_deg:
        info['reason'] = 'already straight'
        return None, info
    if info['after'] > (1.0 - MIN_GAIN) * info['before']:
        info['reason'] = 'no gain'
        return None, info
    if abs(info['params']['rot_deg']) > MAX_ROT_DEG:
        info['reason'] = 'rotation too large'
        return None, info
    corners = np.float64([[inset, inset], [W - inset, inset], [W - inset, H - inset], [inset, H - inset]])
    moved = cv2.perspectiveTransform(corners.reshape(1, -1, 2), Hm).reshape(-1, 2)
    shift = float(np.linalg.norm(moved - corners, axis=1).max())
    info['corner_shift_px'] = round(shift, 1)
    if shift > MAX_CORNER_SHIFT_FRAC * W:
        info['reason'] = 'correction too large'
        return None, info
    info['applied'] = True
    return Hm, info


def apply_refinement(page: np.ndarray, Hm: np.ndarray) -> np.ndarray:
    h, w = page.shape[:2]
    return cv2.warpPerspective(page, Hm, (w, h), flags=cv2.INTER_LINEAR, borderMode=cv2.BORDER_REPLICATE)

File: ports/go/internal/docproc/modules/linerefine.go
package modules

import (
	"math"

	"github.com/protei300/RussianDocsOCR/ports/go/internal/docproc/imaging"
)

// Port of pipeline_modules/page_registration/line_refine.py: straightens a rectified
// page by its own straight structures (LSD segments + projection-profile band tilts),
// fitting a 4-parameter model (rotation, shear, two perspective terms) with a robust
// (Cauchy-weighted IRLS) loss.
//
// Two substitutions from the reference, both forced by the Go OpenCV binding rather
// than chosen for convenience - see their definitions for the measured consequence:
//   - imaging.DetectSegments (Canny+HoughLinesP) stands in for cv2.createLineSegmentDetector,
//     which gocv v0.43.0 does not bind at all.
//   - the 4-parameter fit uses LevenbergMarquardt (this port's own, finite-difference
//     Jacobian) rather than scipy.optimize.least_squares's trust-region algorithm.

const (
	lsdMinLenFrac    = 0.04
	lsdMaxWeight     = 3.0
	angleTolDeg      = 12.0
	profileScale     = 0.5
	profileFineStep  = 0.25
	profileMinPeak   = 1.15
	profileWeight    = 3.0
	minHorizWeight   = 4.0
	maxRotDeg        = 5.0
	maxCornerShiftFr = 0.08
	refineMinGain    = 0.25
	minResidualDeg   = 0.3
	irlsIters        = 3
	irlsScaleDeg     = 1.0
	vertMinLenFrac   = 0.15
	blobMaxFrac      = 0.2
	nBands           = 5
)

var priorScale = [3]float64{0.03, 0.05, 0.03}
var profileCoarse = coarseAngles()

func coarseAngles() []float64 {
	var out []float64
	for a := -6.0; a <= 6.0001; a += 1.0 {
		out = append(out, a)
	}
	return out
}

// measureRow is one row of line_refine.measure()'s (x1,y1,x2,y2,kind,weight,source).
// kind 0 = should be horizontal, 1 = should be vertical.
type measureRow struct {
	X1, Y1, X2, Y2 float64
	Kind           int
	Weight         float64
	Source         int
}

func pymod(a, b float64) float64 {
	m := math.Mod(a, b)
	if m < 0 {
		m += b
	}
	return m
}

func segAngle(x1, y1, x2, y2 float64) float64 {
	ang := math.Atan2(y2-y1, x2-x1) * 180 / math.Pi
	return pymod(ang+90, 180) - 90
}

// otsuInvDropTallBlobs binarises (Otsu, inverted - ink=255) and drops connected
// components taller than blobMaxHeightPx, matching _profile_tilt's/‌_cell_tilts' blob
// filter. Returns the raw row-major bytes (0/255) alongside w, h.
func otsuInvDropTallBlobs(gray imaging.Image, blobMaxHeightPx float64) ([]byte, int, int) {
	binary, _ := imaging.ThresholdOtsu(gray, true)
	defer binary.Close()
	w, h := binary.Width(), binary.Height()
	labels, lw, lh, blobs := imaging.ConnectedComponents8(binary)
	buf, err := binary.Bytes()
	if err != nil || lw != w || lh != h {
		out := make([]byte, w*h)
		copy(out, buf)
		return out, w, h
	}
	out := append([]byte(nil), buf...)
	big := make([]bool, len(blobs))
	for i := 1; i < len(blobs); i++ { // index 0 is background, never dropped
		if float64(blobs[i].H) > blobMaxHeightPx {
			big[i] = true
		}
	}
	anyBig := false
	for _, b := range big {
		anyBig = anyBig || b
	}
	if anyBig {
		for i, lab := range labels {
			if lab >= 0 && int(lab) < len(big) && big[lab] {
				out[i] = 0
			}
		}
	}
	return out, w, h
}

// profileTilt is _profile_tilt (axis=1 only - the only axis line_refine/line_dewarp
// ever use). region is a grayscale crop; returns (tilt degrees, peak ratio); ratio 0
// means "no evidence".
func profileTilt(region imaging.Image) (float64, float64) {
	small := imaging.ResizeArea(region, maxInt(1, int(float64(region.Width())*profileScale)),
		maxInt(1, int(float64(region.Height())*profileScale)))
	defer small.Close()
	w, h := small.Width(), small.Height()
	ink, w2, h2 := otsuInvDropTallBlobs(small, blobMaxFrac*float64(h))
	if w2 != w || h2 != h {
		return 0, 0
	}
	inkImg, err := imaging.NewGrayFromBytes(ink, w, h)
	if err != nil {
		return 0, 0
	}
	defer inkImg.Close()
	validImg := imaging.NewGrayFilled(h, w, 255)
	defer validImg.Close()

	score := func(angles []float64) []float64 {
		out := make([]float64, len(angles))
		for i, a := range angles {
			rot := imaging.RotateNearestZero(inkImg, a)
			cnt := imaging.RotateNearestZero(validImg, a)
			inkSum, errI := imaging.RowSums(rot)
			cntSum, errC := imaging.RowSums(cnt)
			rot.Close()
			cnt.Close()
			if errI != nil || errC != nil {
				out[i] = 0
				continue
			}
			var cmax int64
			for _, c := range cntSum {
				if c > cmax {
					cmax = c
				}
			}
			var prof []float64
			for r := range cntSum {
				if cntSum[r] >= int64(0.6*float64(cmax)) {
					denom := cntSum[r]
					if denom < 1 {
						denom = 1
					}
					prof = append(prof, float64(inkSum[r])/float64(denom))
				}
			}
			if len(prof) > 2 {
				out[i] = varianceF64(prof)
			}
		}
		return out
	}

	coarse := score(profileCoarse)
	ib := argmaxFloat(coarse)
	if ib == 0 || ib == len(profileCoarse)-1 || coarse[ib] <= 0 {
		return 0, 0
	}
	var fine []float64
	for a := profileCoarse[ib] - 1.0; a <= profileCoarse[ib]+1.0001; a += profileFineStep {
		fine = append(fine, a)
	}
	fs := score(fine)
	jb := argmaxFloat(fs)
	best := fine[jb]
	if jb > 0 && jb < len(fine)-1 {
		y0, y1, y2 := fs[jb-1], fs[jb], fs[jb+1]
		den := y0 - 2*y1 + y2
		if den < 0 {
			best += profileFineStep * 0.5 * (y0 - y2) / den
		}
	}
	med := medianOf(coarse)
	ratio := 0.0
	if med > 0 {
		ratio = fs[jb] / med
	}
	return best, ratio
}

func varianceF64(v []float64) float64 {
	var mean float64
	for _, x := range v {
		mean += x
	}
	mean /= float64(len(v))
	var acc float64
	for _, x := range v {
		d := x - mean
		acc += d * d
	}
	return acc / float64(len(v))
}

func medianOf(v []float64) float64 {
	s := append([]float64(nil), v...)
	for i := 1; i < len(s); i++ {
		j := i
		for j > 0 && s[j-1] > s[j] {
			s[j-1], s[j] = s[j], s[j-1]
			j--
		}
	}
	n := len(s)
	if n == 0 {
		return 0
	}
	if n%2 == 1 {
		return s[n/2]
	}
	return 0.5 * (s[n/2-1] + s[n/2])
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// measure is line_refine.measure.
func measure(gray imaging.Image, inset int) []measureRow {
	H, W := float64(gray.Height()), float64(gray.Width())
	var rows []measureRow

	segs := imaging.DetectSegments(gray, lsdMinLenFrac*W)
	for _, s := range segs {
		l := math.Hypot(s.X2-s.X1, s.Y2-s.Y1)
		a := segAngle(s.X1, s.Y1, s.X2, s.Y2)
		w := math.Min(l/(0.1*W), lsdMaxWeight)
		if math.Abs(a) < angleTolDeg {
			rows = append(rows, measureRow{s.X1, s.Y1, s.X2, s.Y2, 0, w, 0})
		} else if math.Abs(math.Abs(a)-90) < angleTolDeg && l >= vertMinLenFrac*H {
			rows = append(rows, measureRow{s.X1, s.Y1, s.X2, s.Y2, 1, w, 0})
		}
	}

	x0, y0 := float64(inset), float64(inset)
	x1, y1 := W-float64(inset), H-float64(inset)
	if x1-x0 < 60 || y1-y0 < 60 {
		return rows
	}
	bh := (y1 - y0) / 3.0
	for _, yb := range linspace(y0, y1-bh, nBands) {
		top, bottom := int(yb), int(yb+bh)
		left, right := int(x0), int(x1)
		if bottom <= top || right <= left {
			continue
		}
		band, err := imaging.ClampedCrop(gray, left, top, right, bottom)
		if err != nil {
			continue
		}
		tilt, ratio := profileTilt(band)
		band.Close()
		if ratio >= profileMinPeak {
			yc := yb + 0.5*bh
			half := 0.5 * (x1 - x0)
			dy := math.Tan(tilt*math.Pi/180) * half
			w := profileWeight * math.Min(1.0, ratio-1.0)
			rows = append(rows, measureRow{x0, yc - dy, x1, yc + dy, 0, w, 1})
		}
	}
	return rows
}

// refineModel is line_refine._model.
func refineModel(params [4]float64, W, H float64) [3][3]float64 {
	th, s, p1, p2 := params[0], params[1], params[2], params[3]
	k := 2.0 / W
	T := [3][3]float64{{k, 0, -1}, {0, k, -H / W}, {0, 0, 1}}
	R := [3][3]float64{{math.Cos(th), -math.Sin(th), 0}, {math.Sin(th), math.Cos(th), 0}, {0, 0, 1}}
	S := [3][3]float64{{1, s, 0}, {0, 1, 0}, {0, 0, 1}}
	P := [3][3]float64{{1, 0, 0}, {0, 1, 0}, {p1, p2, 1}}
	invT, ok := imaging.InvertH(T)
	if !ok {
		return imaging.Identity3()
	}
	return imaging.MulH(invT, imaging.MulH(P, imaging.MulH(R, imaging.MulH(S, T))))
}

// anglesAfter is line_refine._angles_after.
func anglesAfter(Hm [3][3]float64, meas []measureRow) []float64 {
	out := make([]float64, len(meas))
	for i, m := range meas {
		p0 := imaging.TransformPoint(Hm, imaging.Point{X: m.X1, Y: m.Y1})
		p1 := imaging.TransformPoint(Hm, imaging.Point{X: m.X2, Y: m.Y2})
		ang := math.Atan2(p1.Y-p0.Y, p1.X-p0.X) * 180 / math.Pi
		ang = pymod(ang+90, 180) - 90
		if m.Kind == 0 {
			out[i] = ang
		} else {
			out[i] = math.Abs(ang) - 90
		}
	}
	return out
}

func refineResiduals(params []float64, meas []measureRow, W, H float64, wgt []float64) []float64 {
	var p4 [4]float64
	copy(p4[:], params)
	dev := anglesAfter(refineModel(p4, W, H), meas)
	out := make([]float64, 0, len(meas)+3)
	for i, d := range dev {
		out = append(out, d*math.Sqrt(wgt[i]))
	}
	out = append(out, params[1]/priorScale[0], params[2]/priorScale[1], params[3]/priorScale[2])
	return out
}

func fitRefineParams(meas []measureRow, W, H float64) [4]float64 {
	x := []float64{0, 0, 0, 0}
	wgt := make([]float64, len(meas))
	for i, m := range meas {
		wgt[i] = m.Weight
	}
	for iter := 0; iter < irlsIters; iter++ {
		wgtCopy := append([]float64(nil), wgt...)
		x = LevenbergMarquardt(x, func(p []float64) []float64 {
			return refineResiduals(p, meas, W, H, wgtCopy)
		}, 100)
		var p4 [4]float64
		copy(p4[:], x)
		r := anglesAfter(refineModel(p4, W, H), meas)
		for i, m := range meas {
			wgt[i] = m.Weight / (1.0 + (r[i]/irlsScaleDeg)*(r[i]/irlsScaleDeg))
		}
	}
	var out [4]float64
	copy(out[:], x)
	return out
}

// StraightenInfo is diagnostic only (not part of the conformance contract - no stage
// reads it); kept small on purpose.
type StraightenInfo struct {
	Applied bool
	Reason  string
}

// RefineByLines is line_refine.refine_by_lines: the homography that straightens a
// rectified page by its own lines, or nil when there is not enough evidence or the
// correction is not warranted.
func RefineByLines(gray imaging.Image, inset int) (*[3][3]float64, StraightenInfo) {
	H, W := float64(gray.Height()), float64(gray.Width())
	meas := measure(gray, inset)
	if len(meas) == 0 {
		return nil, StraightenInfo{Reason: "no evidence"}
	}
	horizW := 0.0
	for _, m := range meas {
		if m.Kind == 0 {
			horizW += m.Weight
		}
	}
	if horizW < minHorizWeight {
		return nil, StraightenInfo{Reason: "too little horizontal evidence"}
	}
	weights := make([]float64, len(meas))
	for i, m := range meas {
		weights[i] = m.Weight
	}
	before0 := anglesAfter(imaging.Identity3(), meas)
	beforeAbs := make([]float64, len(before0))
	for i, v := range before0 {
		beforeAbs[i] = math.Abs(v)
	}
	before := wmedian(beforeAbs, weights)

	x := fitRefineParams(meas, W, H)
	Hm := refineModel(x, W, H)
	after0 := anglesAfter(Hm, meas)
	afterAbs := make([]float64, len(after0))
	for i, v := range after0 {
		afterAbs[i] = math.Abs(v)
	}
	after := wmedian(afterAbs, weights)

	if before < minResidualDeg {
		return nil, StraightenInfo{Reason: "already straight"}
	}
	if after > (1.0-refineMinGain)*before {
		return nil, StraightenInfo{Reason: "no gain"}
	}
	rotDeg := x[0] * 180 / math.Pi
	if math.Abs(rotDeg) > maxRotDeg {
		return nil, StraightenInfo{Reason: "rotation too large"}
	}
	corners := []imaging.Point{
		{X: float64(inset), Y: float64(inset)}, {X: W - float64(inset), Y: float64(inset)},
		{X: W - float64(inset), Y: H - float64(inset)}, {X: float64(inset), Y: H - float64(inset)},
	}
	moved := imaging.TransformPoints(Hm, corners)
	shift := 0.0
	for i := range corners {
		d := dist(corners[i], moved[i])
		if d > shift {
			shift = d
		}
	}
	if shift > maxCornerShiftFr*W {
		return nil, StraightenInfo{Reason: "correction too large"}
	}
	return &Hm, StraightenInfo{Applied: true}
}

// ApplyRefinement warps a page through the straightening homography, matching
// line_refine.apply_refinement (linear interpolation, replicate border).
func ApplyRefinement(page imaging.Image, Hm [3][3]float64) imaging.Image {
	return imaging.WarpByHomography(page, Hm, page.Width(), page.Height(), true)
}
